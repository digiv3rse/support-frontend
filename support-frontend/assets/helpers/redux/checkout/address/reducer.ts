import type { PayloadAction } from '@reduxjs/toolkit';
import { combineReducers, createSlice } from '@reduxjs/toolkit';
import { postcodeFinderReducerFor } from 'components/subscriptionCheckouts/address/postcodeFinderStore';
import { fromString } from 'helpers/internationalisation/country';
import type { AddressType } from 'helpers/subscriptionsForms/addressType';
import { removeError } from 'helpers/subscriptionsForms/validation';
import type { AddressFormFieldError } from './state';
import { getInitialAddressFieldsState } from './state';

function getAddressSlice(type: AddressType) {
	const fieldsSlice = createSlice({
		name: `${type}AddressFields`,
		initialState: getInitialAddressFieldsState,
		reducers: {
			setLineOne(state, action: PayloadAction<string>) {
				state.lineOne = action.payload;
				state.errors = removeError('lineOne', state.errors);
			},
			setLineTwo(state, action: PayloadAction<string>) {
				state.lineTwo = action.payload;
				state.errors = removeError('lineTwo', state.errors);
			},
			setTownCity(state, action: PayloadAction<string>) {
				state.city = action.payload;
				state.errors = removeError('city', state.errors);
			},
			setState(state, action: PayloadAction<string>) {
				state.state = action.payload;
				state.errors = removeError('state', state.errors);
			},
			setPostcode(state, action: PayloadAction<string>) {
				state.postCode = action.payload;
				state.errors = removeError('postCode', state.errors);
			},
			setCountry(state, action: PayloadAction<string>) {
				const country = fromString(action.payload);

				if (country) {
					state.country = country;
				}
			},
			setFormErrors(state, action: PayloadAction<AddressFormFieldError[]>) {
				state.errors = action.payload;
			},
		},
	});
	const fieldsReducer = fieldsSlice.reducer;

	const reducer = combineReducers({
		fields: fieldsReducer,
		postcode: postcodeFinderReducerFor(type),
	});

	return { fieldsSlice, reducer };
}

export const {
	fieldsSlice: deliveryAddressFieldsSlice,
	reducer: deliveryAddressReducer,
} = getAddressSlice('delivery');

export const {
	fieldsSlice: billingAddressFieldsSlice,
	reducer: billingAddressReducer,
} = getAddressSlice('billing');
