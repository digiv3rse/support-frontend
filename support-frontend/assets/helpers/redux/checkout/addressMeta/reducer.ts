import type { PayloadAction } from '@reduxjs/toolkit';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import type { DeliveryAgentsResponse } from './state';
import { initialState } from './state';

export const addressMetaSlice = createSlice({
	name: 'addressMeta',
	initialState,
	reducers: {
		setBillingAddressMatchesDelivery(state, action: PayloadAction<boolean>) {
			state.billingAddressMatchesDelivery = action.payload;
		},
		setDeliveryInstructions(state, action: PayloadAction<string>) {
			state.deliveryInstructions = action.payload;
		},
		setDeliveryAgent(state, action: PayloadAction<number | undefined>) {
			state.deliveryAgent.chosenAgent = action.payload;
		},
		setDeliveryAgentResponse(
			state,
			action: PayloadAction<DeliveryAgentsResponse | undefined>,
		) {
			state.deliveryAgent.response = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(getDeliveryAgentsThunk.pending, (state) => {
			state.deliveryAgent.isLoading = true;
			state.deliveryAgent.error = undefined;
		});
		builder.addCase(
			getDeliveryAgentsThunk.fulfilled,
			(state, action: PayloadAction<DeliveryAgentsResponse>) => {
				state.deliveryAgent.isLoading = false;
				state.deliveryAgent.response = action.payload;
			},
		);
		builder.addCase(getDeliveryAgentsThunk.rejected, (state, action) => {
			state.deliveryAgent.isLoading = false;
			state.deliveryAgent.error = action.error.message;
		});
	},
});

export const getDeliveryAgentsThunk = createAsyncThunk(
	`addressMeta/getDeliveryAgents`,
	getDeliveryAgents,
);

async function getDeliveryAgents(postcode: string) {
	const agentsResponse = await fetch(`/delivery-agents/${postcode}`);
	return agentsResponse.json();
}

export const addressMetaReducer = addressMetaSlice.reducer;
