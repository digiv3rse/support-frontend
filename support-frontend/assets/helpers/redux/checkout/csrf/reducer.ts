import { createSlice } from '@reduxjs/toolkit';
import { initialCsrfState } from './state';

export const csrfSlice = createSlice({
	name: 'csrf',
	initialState: initialCsrfState,
	reducers: {},
});

export const csrfReducer = csrfSlice.reducer;
