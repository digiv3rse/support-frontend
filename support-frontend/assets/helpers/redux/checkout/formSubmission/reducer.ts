import { createSlice } from '@reduxjs/toolkit';
import { initialState } from './state';
import { submitForm } from './thunks';

export const formSubmissionSlice = createSlice({
	name: 'formSubmission',
	initialState,
	reducers: {},
	extraReducers: (builder) => {
		builder.addCase(submitForm.fulfilled, (state, action) => {
			state.status = action.payload;
		});

		builder.addCase(submitForm.pending, (state) => {
			state.status = 'pending';
		});

		builder.addCase(submitForm.rejected, (state, action) => {
			state.status = 'error';
			state.error = action.meta.errorMessage;
		});
	},
});

export const formSubmissionReducer = formSubmissionSlice.reducer;
