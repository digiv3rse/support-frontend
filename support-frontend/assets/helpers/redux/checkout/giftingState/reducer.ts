import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';
import type { Title } from 'helpers/user/details';
import { initialGiftingState } from './state';

export const giftingSlice = createSlice({
	name: 'gifting',
	initialState: initialGiftingState,
	reducers: {
		setTitle(state, action: PayloadAction<Title>) {
			const newTitle =
				action.payload !== 'Select a title' ? action.payload : state.title;
			state.title = newTitle;
		},
		setFirstName(state, action: PayloadAction<string>) {
			state.firstName = action.payload;
		},
		setLastName(state, action: PayloadAction<string>) {
			state.lastName = action.payload;
		},
		setEmail(state, action: PayloadAction<string>) {
			state.email = action.payload;
		},
		setGiftMessage(state, action: PayloadAction<string>) {
			state.giftMessage = action.payload;
		},
		setGiftDeliveryDate(state, action: PayloadAction<string>) {
			state.giftDeliveryDate = action.payload;
		},
	},
});

export const giftingReducer = giftingSlice.reducer;
