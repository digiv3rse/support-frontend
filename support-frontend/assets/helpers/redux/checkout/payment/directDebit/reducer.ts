import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';
import { getSliceErrorsFromZodResult } from 'helpers/redux/utils/validation/errors';
import { validateForm } from '../../checkoutActions';
import type { Phase, SortCodeUpdate } from './state';
import { directDebitSchema, initialDirectDebitState } from './state';
import {
	confirmAccountDetails,
	directDebitErrorMessages,
	payWithDirectDebit,
} from './thunks';

export const directDebitSlice = createSlice({
	name: 'directDebit',
	initialState: initialDirectDebitState,
	reducers: {
		setPopupOpen(state) {
			state.isPopUpOpen = true;
		},
		setPopupClose(state) {
			state.isPopUpOpen = false;
		},
		setDDGuaranteeOpen(state) {
			state.isDDGuaranteeOpen = true;
		},
		setDDGuaranteeClose(state) {
			state.isDDGuaranteeOpen = false;
		},
		setSortCode(state, action: PayloadAction<SortCodeUpdate>) {
			const { index, partialSortCode } = action.payload;
			state.sortCodeArray.splice(index, 1, partialSortCode);
		},
		setSortCodeString(state, action: PayloadAction<string>) {
			state.sortCodeString = action.payload;
		},
		setAccountNumber(state, action: PayloadAction<string>) {
			state.accountNumber = action.payload;
		},
		setAccountHolderName(state, action: PayloadAction<string>) {
			state.accountHolderName = action.payload;
		},
		setAccountHolderConfirmation(state, action: PayloadAction<boolean>) {
			state.accountHolderConfirmation = action.payload;
		},
		setFormError(state, action: PayloadAction<string>) {
			state.formError = action.payload;
		},
		resetFormError(state) {
			state.formError = '';
		},
		setPhase(state, action: PayloadAction<Phase>) {
			state.phase = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(confirmAccountDetails.pending, (state) => {
			state.formError = '';
		});

		builder.addCase(confirmAccountDetails.fulfilled, (state, action) => {
			if (action.payload) {
				state.phase = 'confirmation';
			} else {
				state.formError = directDebitErrorMessages.incorrectInput;
			}
		});

		builder.addCase(confirmAccountDetails.rejected, (state, action) => {
			const errorMessage =
				action.error.message ?? directDebitErrorMessages.default;
			state.formError = errorMessage;
		});

		builder.addCase(payWithDirectDebit.fulfilled, (state) => {
			state.isPopUpOpen = false;
		});

		builder.addCase(payWithDirectDebit.rejected, (state) => {
			state.isPopUpOpen = false;
		});

		builder.addCase(validateForm, (state) => {
			const validationResult = directDebitSchema.safeParse(state);

			if (!validationResult.success) {
				state.errors = getSliceErrorsFromZodResult(
					validationResult.error.format(),
				);
			}
		});
	},
});

export const directDebitReducer = directDebitSlice.reducer;
