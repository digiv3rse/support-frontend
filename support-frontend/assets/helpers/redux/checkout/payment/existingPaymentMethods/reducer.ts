import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';
import { Completed, Failed, Pending } from 'helpers/types/asyncStatus';
import { logException } from 'helpers/utilities/logger';
import { setProductType } from '../../product/actions';
import type { RecentlySignedInExistingPaymentMethod } from './state';
import { initialState } from './state';
import { getExistingPaymentMethods } from './thunks';
import { getUsableExistingPaymentMethods } from './utils';

export const existingPaymentMethodsSlice = createSlice({
	name: 'existingPaymentMethods',
	initialState,
	reducers: {
		selectExistingPaymentMethod(
			state,
			action: PayloadAction<RecentlySignedInExistingPaymentMethod>,
		) {
			state.selectedPaymentMethod = action.payload;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(setProductType, (state, action) => {
			if (action.payload === 'ONE_OFF') {
				state.showExistingPaymentMethods = false;
			} else {
				state.showExistingPaymentMethods = true;
			}
		});

		builder.addCase(getExistingPaymentMethods.pending, (state) => {
			state.status = Pending;
		});

		builder.addCase(getExistingPaymentMethods.rejected, (state, action) => {
			logException(
				'Failed to get existing payment options',
				action.error as Record<string, unknown>,
			);
			state.status = Failed;
		});

		builder.addCase(getExistingPaymentMethods.fulfilled, (state, action) => {
			const usablePaymentMethods = getUsableExistingPaymentMethods(
				action.payload,
			);
			state.paymentMethods = usablePaymentMethods;
			state.status = Completed;

			// The user has at least one existing payment method, but it's not currently usable for payment
			if (usablePaymentMethods.length < action.payload.length) {
				state.showReauthenticateLink = true;
			}
		});
	},
});

export const existingPaymentMethodsReducer =
	existingPaymentMethodsSlice.reducer;
