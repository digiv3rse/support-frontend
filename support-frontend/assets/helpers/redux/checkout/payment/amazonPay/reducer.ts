import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';
import { validateForm } from '../../checkoutActions';
import { submitForm } from '../../formSubmission/thunks';
import { initialAmazonPayState } from './state';

export const amazonPaySlice = createSlice({
	name: 'amazonPay',
	initialState: initialAmazonPayState,
	reducers: {
		setAmazonPayWalletIsStale(state, action: PayloadAction<boolean>) {
			state.walletIsStale = action.payload;
		},
		setAmazonPayHasAccessToken(state) {
			state.hasAccessToken = true;
		},
		setAmazonPayFatalError(state) {
			state.fatalError = true;
		},
		setAmazonPayPaymentSelected(state, action: PayloadAction<boolean>) {
			state.paymentSelected = action.payload;
			delete state.errors.paymentSelected;
		},
		setAmazonPayOrderReferenceId(state, action: PayloadAction<string>) {
			state.orderReferenceId = action.payload;
		},
		setAmazonPayBillingAgreementId(state, action: PayloadAction<string>) {
			state.amazonBillingAgreementId = action.payload;
		},
		setAmazonPayBillingAgreementConsentStatus(
			state,
			action: PayloadAction<boolean>,
		) {
			state.amazonBillingAgreementConsentStatus = action.payload;
			delete state.errors.consentStatus;
		},
	},
	extraReducers: (builder) => {
		builder.addCase(validateForm, (state, action) => {
			if (action.payload === 'AmazonPay') {
				if (!state.paymentSelected) {
					state.errors.paymentSelected = ['Please proceed with Amazon Pay'];
				}
				if (!state.amazonBillingAgreementConsentStatus) {
					state.errors.consentStatus = [
						'Please tick the box to agree to a recurring payment',
					];
				}
			}
		});

		builder.addCase(submitForm.rejected, (state, action) => {
			const { paymentMethod, errorMessage } = action.meta;
			if (paymentMethod === 'AmazonPay') {
				if (
					errorMessage === 'amazon_pay_try_other_card' ||
					errorMessage === 'amazon_pay_try_again'
				) {
					// Must re-render the wallet widget in order to display amazon's error message
					state.walletIsStale = true;
				} else {
					// Disable Amazon Pay
					state.fatalError = true;
				}
			}
		});
	},
});

export const amazonPayReducer = amazonPaySlice.reducer;
