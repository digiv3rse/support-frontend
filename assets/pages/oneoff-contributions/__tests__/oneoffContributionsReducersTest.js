// @flow
import reducer from '../oneOffContributionsReducers';


describe('One-off Reducer', () => {

  it('should return the initial state', () => {
    expect(reducer(20, 'GBP')(undefined, {})).toMatchSnapshot();
  });

  it('should handle CHECKOUT_ERROR', () => {

    const message = 'Test error';
    const action = {
      type: 'CHECKOUT_ERROR',
      message,
    };

    const newState = reducer(20, 'GBP')(undefined, action);

    expect(newState.oneoffContrib.error).toEqual(message);
  });

  it('should handle SET_PAYPAL_BUTTON', () => {

    const value = 'ExpressCheckout';
    const action = {
      type: 'SET_PAYPAL_BUTTON',
      value,
    };

    const newState = reducer(20, 'GBP')(undefined, action);

    expect(newState.oneoffContrib.error).toMatchSnapshot();
  });

});
