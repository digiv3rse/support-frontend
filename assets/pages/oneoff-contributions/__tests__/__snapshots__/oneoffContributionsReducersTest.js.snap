// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`One-off Reducer should handle SET_PAYPAL_BUTTON 1`] = `null`;

exports[`One-off Reducer should return the initial state 1`] = `
Object {
  "checkoutForm": Object {
    "email": Object {
      "pattern": /\\^\\[a-zA-Z0-9\\.!#\\$%&'\\*\\+\\\\/=\\?\\^_\`\\{\\|\\}~-\\]\\+@\\[a-zA-Z0-9\\]\\(\\?:\\[a-zA-Z0-9-\\]\\{0,61\\}\\[a-zA-Z0-9\\]\\)\\?\\(\\?:\\\\\\.\\[a-zA-Z0-9\\]\\(\\?:\\[a-zA-Z0-9-\\]\\{0,61\\}\\[a-zA-Z0-9\\]\\)\\?\\)\\+\\$/,
      "required": true,
      "shouldValidate": false,
    },
    "fullName": Object {
      "pattern": /\\.\\*/,
      "required": true,
      "shouldValidate": false,
    },
  },
  "csrf": Object {
    "token": null,
  },
  "marketingConsent": Object {
    "confirmOptIn": null,
    "error": false,
  },
  "oneoffContrib": Object {
    "amount": 20,
    "error": null,
    "paymentComplete": false,
  },
  "user": Object {
    "displayName": "",
    "email": "",
    "firstName": "",
    "fullName": "",
    "gnmMarketing": false,
    "id": "",
    "isPostDeploymentTestUser": false,
    "isSignedIn": false,
    "isTestUser": null,
    "lastName": "",
  },
}
`;
